Lock/Unlock throttle offending files

fjvallarino opened the pull request:
@tejohnso please review

tejohnso commented:
The lock down handler shouldn't enqueue a task if the file is already throttled.

And I think the marking of RvStorageObject with status=throttled/not_throttled should be done by the remove public read server task and add public read server tasks once they have completed successfully.

So maybe the get throttled files method and the methods to set various statuses should be in the rvstorageobjectAccessor.  Those are methods that need to be shared across several classes.

fjvallarino commented:
@tejohnso Updated

tejohnso commented on src/main/java/com/risevision/storage/api/accessors/RvStorageObjectAccessor.java:
```diff
@@ -146,11 +147,11 @@ else if(type == TagType.FREEFORM) {
+  public RvStorageObject get(String companyId, String objectId) {
     List<RvStorageObject> list = datastoreService.list(RvStorageObject.class, "companyId", companyId, "objectId", objectId);
     
     if(list.size() > 0) {
```
I think a severe error should be thrown here.

tejohnso commented on src/main/java/com/risevision/storage/api/accessors/RvStorageObjectAccessor.java:
```diff
@@ -164,6 +165,27 @@ public RvStorageObject get(String companyId, String objectId) throws Exception {
+      rvso = new RvStorageObject(companyId, objectId);
+    }
+    
+    if(throttle) {
```
Can we get rid of boolean throttle and instead check for offenders is or is not null?

fjvallarino replied:
I'm using the boolean because I think RemovePublicRead can be invoked from other places which are not providing the list of offenders.

tejohnso commented on src/main/java/com/risevision/storage/api/accessors/RvStorageObjectAccessor.java:
```diff
@@ -250,6 +272,13 @@ else if(tagType == TagType.FREEFORM) {
+            RvStorageObject.THROTTLE_STATUS.NEEDS_UNTHROTTLING));
+  }
+  
   protected Map<String, TagDefinition> getTagDefinitions(String companyId) throws Exception {
```
We moved away from the "NEEDS_" statuses, right?

tejohnso commented on src/main/java/com/risevision/storage/servertasks/AddPublicReadObjectServerTask.java:
```diff
@@ -34,6 +38,8 @@ public void handleRequest() throws IOException {
       .execute();
+      
+      new RvStorageObjectAccessor().updateThrottleStatus(Globals.COMPANY_BUCKET_PREFIX.replace(Globals.COMPANY_BUCKET_PREFIX, ""), object, false, null);
     } catch (GoogleJsonResponseException e) {
```
The call to updateThrottleStatus is missing the bucket isn't it?
And above that (line39), I think the requestParams.get("") calls can be changed to use the new bucket and object variables.

fjvallarino commented:
@tejohnso Updated to reflect the new PublicReadStatus

tejohnso commented on src/test/java/unit/com/risevision/storage/servertasks/RemovePublicReadObjectServerTaskTest.java:
```diff
@@ -12,7 +15,7 @@
 import com.google.api.services.storage.model.ObjectAccessControl;
 
-public class RemovePublicReadObjectServerTaskTest {
+public class RemovePublicReadObjectServerTaskTest extends ObjectifyTest {
```
Is something else missing here?

fjvallarino replied:
The tests could check the status was updated accordingly. Extending from ObjectifyTest is mandatory since now we're making a database call which otherwise fails.

tejohnso commented on src/main/java/com/risevision/storage/servertasks/UpdateThrottlingLockHandler.java:
```diff
@@ -0,0 +1,67 @@
+    RvStorageObjectAccessor storageObjectAccessor = new RvStorageObjectAccessor();
+    List<RvStorageObject> throttledFiles = storageObjectAccessor.getThrottledFiles();
+    Map<String, RvStorageObject> throttledFilesMap = buildFilesMap(throttledFiles);
+    Map<String, RvStorageObject> offendingFilesMap = buildFilesMap(rp.getRvStorageObjects());
```
The offendingFilesMap is already in the RowProcessor.  It can be taken from there rather than build rebuilt.  
And maybe storageObjectAccessor.getThrottledFiles() can return a map instead of a List.  Then the caller has the map, and if they want a list they can just do map.values()

tejohnso commented on src/main/java/com/risevision/storage/servertasks/UpdateThrottlingLockHandler.java:
```diff
@@ -0,0 +1,67 @@
+      if(!offendingFilesMap.containsKey(rvso.getId())) {
+        storageObjectAccessor.updateThrottleOffenders(rvso.getCompanyId(), rvso.getObjectId(), null);
+        
+        enqueueAddPublicRead(rvso.getCompanyId(), rvso.getObjectId(), PublicReadStatus.THROTTLED);
```
After updating the throttle offenders here, we should check to see if the publicReadStatus is of length one and contains THROTTLED.  If so, only then enqueue the addPublicRead.  Otherwise just remove the THROTTLED status.

Similar for removals.

tejohnso commented on src/main/java/com/risevision/storage/servertasks/ThrottleOffenderRowProcessor.java:
```diff
@@ -18,7 +16,7 @@
   private HashMap<String, RvStorageObject> rvStorageObjectsMap;
```
I don't think the OffendersRowProcessor needs to map the entire RvStorageObject.  Maybe <String, List<ThrottleOffender>> would suffice.   And there should be a getOffendersMap()

fjvallarino replied:
I think there's a missing part in this comment (the "Maybe >" part).

fjvallarino commented:
All the items were addressed, except this one: *I don't think the OffendersRowProcessor needs to map the entire RvStorageObject. Maybe String, List[ThrottleOffender] would suffice. And there should be a getOffendersMap()*

@tejohnso please review

tejohnso commented on src/main/java/com/risevision/storage/servertasks/AddPublicReadObjectServerTask.java:
```diff
@@ -31,9 +38,13 @@ public void handleRequest() throws IOException {
-      .execute();
+      RvStorageObject storageObject = new RvStorageObjectAccessor().removePublicReadStatus(bucket.replace(Globals.COMPANY_BUCKET_PREFIX, ""), object, reason);
+      
+      if(storageObject.isPublicRead()) {
```
There should be a negation here, right?

fjvallarino replied:
No, because that means the Datastore object does not have any remaining reasons for being blocked. Maybe it could be changed to see if only one reason is remaning (and it is the one we're currently removing), and only then removePublicReadStatus + insert acl.

tejohnso replied:
I think we need to clarify the difference between the actual public read status, and the public read status indicated on the rvso.

Maybe we should change the variable to publicReadProhibitReason[NOT_PROHIBITED, INACTIVE_SUBSCRIPTION, THROTTLED]

Then have
add/removePublicReadProhibitReason()

fjvallarino replied:
So, should we make sure non-prohibited files always have the NOT_PROHIBITED flag? We could use that to fetch all files which do not have public read removed, but that also means having to keep a record for all the files in Storage.

tejohnso replied:
I think we should set NOT_PROHIBITED when the accessor has cleared out any prohibit reasons.  So the field reflects the reality that it isn't prohibited.   The option of adding NOT_PROHIBITED to all files is something we may use in the future.

tejohnso commented on src/main/java/com/risevision/storage/servertasks/RemovePublicReadObjectServerTask.java:
```diff
@@ -14,13 +17,22 @@
+    RvStorageObject storageObject = new RvStorageObjectAccessor().addPublicReadStatus(bucket.replace(Globals.COMPANY_BUCKET_PREFIX, ""), object, reason);
+    
+    // If only one item exists, it's the one we just added
+    if(storageObject.getPublicReadStatus().size() == 1) {
```
If storageObject.getPublicReadStatus() was just [PUBLIC] then size would now be 2 and we wouldn't update the acl.

fjvallarino replied:
We're never actually inserting PUBLIC, so that should not happen.

tejohnso replied:
PUBLIC is part of the enum so we shouldnt expect it not to happen.   Plus....i think we need to set PUBLIC when it is.    Although I think we should change the variable name around as mentioned in my other comment.

tejohnso commented on src/main/java/com/risevision/storage/servertasks/UpdateThrottlingLockHandler.java:
```diff
@@ -0,0 +1,72 @@
+      
+      rvso = storageObjectAccessor.updateThrottleOffenders(rvso.getCompanyId(), rvso.getObjectId(), rvso.getThrottleOffenders());
+      
+      if(rvso.getPublicReadStatus() != null && rvso.getPublicReadStatus().size() == 0 && (throttled == null || !throttled.isThrottled())) {
```
If getPublicReadStatus() is null we still need to remove public read, correct?

tejohnso commented:
@fjvallarino I added a couple of small commits.  They don't address previous comments.  One is just for column width, the other is to remove that map build method.

fjvallarino commented:
Renamed PublicReadStatus to PublicReadProhibitReason (plus PUBLIC to NOT_PROHIBITED). Updated all the getters and related functions. Since now NOT_PROHIBITED is added when no reasons remain, the test cases were updated to reflect this.

@tejohnso please review

tejohnso commented:
:+1:

tejohnso commented:
@fjvallarino Please review the temporary change.  If possible we should merge this as is and then uncomment the public read removal line once the UI cues are in place.  

We'll also need to revisit the enablePublicRead process that occurs automatically whenever a user signs in to the client.  We don't want to unthrottle their files just because they have a subscription.

fjvallarino commented:
:+1:
